Maelstrom/AbilityShuffle.cs

```diff
@@ -23,82 +23,123 @@ public static List<JunctionableGF> Randomise(FileSource mainSource, int seed, St
-                    init.GFs[i].Abilities = new BitArray(init.GFs[i].Abilities.Length, false);
+                    if (learnSlotIndex < 4 && settings.GfAbilitiesBasics)
+                    {
+                        AddBasicAbilities(kernel.JunctionableGFs[gfId].Abilities, init, gfId, unusedAbilities);
 
-                    var unusedAbilities = includedAbilities.Select(a => a.AbilityID).ToList();
-                    for (int j = 0; j < 21; j++)
+                        learnSlotIndex = 3;
+                    }
+                    else
                     {
-                        // guarantee basic commands (magic, gf, draw, item)
-                        if (j < 4 && settings.GfAbilitiesBasics)
-                        {
-                            var id = j + 20;
-                            kernel.JunctionableGFs[i].Abilities[j].Ability = (byte)id;
-                            kernel.JunctionableGFs[i].Abilities[j].Unlocker = 0;
-                            init.GFs[i].Abilities[id] = true;
-                            unusedAbilities.Remove(id);
-                            continue;
-                        }
-
-                        // roll abilities
-                        var ability = unusedAbilities[random.Next(unusedAbilities.Count)];
-                        kernel.JunctionableGFs[i].Abilities[j].Ability = (byte)ability;
-                        kernel.JunctionableGFs[i].Abilities[j].Unlocker = 0;
-                        init.GFs[i].Abilities[ability] = (ability >= 20 && ability <= 23);
-                        unusedAbilities.Remove(ability);
+                        AddRandomAbility(random, kernel.JunctionableGFs[gfId].Abilities, learnSlotIndex, init.GFs[gfId], unusedAbilities);
                     }
+                }
 
-                    // sort abilities
-                    kernel.JunctionableGFs[i].Abilities = kernel.JunctionableGFs[i].Abilities.OrderBy(a => a.Ability == 0 ? byte.MaxValue : a.Ability).ToArray();
-
-                    // clear ability being learned
-                    init.GFs[i].CurrentAbility = 0;
+                //This duplicates the first ability for the remaining slots. It's hacky, but it works.
+                for (int abilityIndex = abilityLimit; abilityIndex < 21; abilityIndex++)
+                {
+                    var firstAbility = kernel.JunctionableGFs[gfId].Abilities[0].Ability;
+                    var firstLearned = init.GFs[gfId].Abilities[firstAbility];
+                    AddAbility(kernel.JunctionableGFs[gfId].Abilities, abilityIndex, init.GFs[gfId], firstAbility, firstLearned);
                 }
+
+                // clear ability being learned
+                init.GFs[gfId].CurrentAbility = 0;
             }
+        }
 
-            // reduce "empty" cost
-            kernel.Abilities[24].APCost = 60;
+        private static void AddBasicAbilities(GFAbility[] abilities, Init init, int gfId, List<int> unusedAbilities)
```

Comment:
GF abilities are now in a list instead of an array - if you make `abilities` an `IList<GFAbility>` it can accept either type

Reply:
Cool. There's a bit of refactoring I want to do to all of this anyway. I'm passing around GFAbility[] and Init like a crazy person.